### ISSUE TITLE
Document how to run ratbag with non-root account

### ISSUE THREAD

**Issue by hipnoizz** [libratbag/libratbag#679]
I cloned the repo, built the project according to https://github.com/libratbag/libratbag/blob/master/README.md. The ratbag was installed into `/usr/local`. Then I've tried to run ratbagd by going into `/usr/local/bin/` and typing `./ratbagd`. I've got _ratbagd error: Failed to start ratbagd: Permission denied_ so I've tried `sudo ./ratbagd`. No cookies as well, the same error.

Now, I've managed to run `sudo ./ratbagctl.devel` from `builddir` and it was able to list and configure a SteelSeries Sensei Raw mouse which is at the moment supported. But what I really would like to achieve is possibility to run the app from IDE (I've installed Gnome Builder - last time I wrote in C language was like 20 years ago, so I don't know any better). When I try do that I got familiar _ratbagd error: Failed to start ratbagd: Permission denied_. The error happens in `sd_bus_request_name`. https://github.com/libratbag/libratbag/wiki/Developers mentions that it is possible to run everything as non-root user but unfortunately it isn't documented anywhere.

Overall I wanted to add support for Roccat Kone Pure (Black Edition specifically, there are like 10 various Roccat Kone Pure variants, I have not idea if their protocol differ...). I installed Windows 7, then installed SteelSeries software, got Wireshark sniffing on USB connection and after some time I had pretty good understanding what data/packets are exchanged. Then I've looked at the Kone XTD driver to see how the drivers are structured. Since it will be some experimenting (can someone bricks his mouse?) the easiest way to do that, at least for me, is to run the app from the IDE, get a feeling about an overall code flow, see what bytes are exchanged, introspect variables etc. etc. I have no idea how to do that right now.

It's just my gut feeling, but I think that there may be like three types of developers interested in extending ratbag:
- those that want to add support for their mice,
- those that want to extend ratbag capabilities (e.g. the Sensei Raw mouse has additional type of LED mode, when the LED flashes when a button is clicked),
- those developing ratbag internals - (USB HID protocol, support for new high-level capabilities, general architecture, gritty details like error handling).

Obviously I belong to the first group and what I'm really interested in is how to write a new driver or extend an existing `roccat` driver. I'm not interested what D-Bus is, and how it interacts with systemd, ratbag internals are not my concern really besides what drivers API is provided. It would be great to have docs that strive to remove any friction for this kind of devs as well - guys with an experience with Linux/Gnome/D-Bus/Meson stuff will manage on their own anyway and don't need much hand-holding. I browsed through the issues and I can see at least several places when newcomers struggled with various aspects of the driver development and judging from a lack of drivers for the devices they owned my assumptions is that they got disheartened at some point.

I realize that writing docs are no fun (especially if no one is paying for it) - I don't have experience with C/Gnome development but nonetheless I'm a software developers for over 15 years so I understand you. But my take is that removing obstacles for wannabe drivers developers would allow ratbag to support more and more devices (and have more bugs as well ;)).

**Comment by FFY00** [comment]
Sorry for the misclick.

I agree that we should document this. Either way, this shouldn't limit in any way your ability to develop for libratbag. I didn't really understand your issue, do you want to test the changes in an IDE environment, is that it?

**Comment by hipnoizz** [comment]
Yes, that would be ideal - make changes, build the project, run `ratbagd` in the debug mode - everything from IDE - then e.g. use `ratbagctl` to interact with the daemon, with IDE providing UI for setting breakpoints and stepping through the lines (assuming IDE can integrate with gdb - Gnome Builder seems to do that) as I'm playing with the driver code. There may be simpler ways, but for someone not accustomed with Linux development this one is the most obvious.

From what I collected from various bits and pieces there are at least 2 problems when running as non-root:
- access to hidraw - https://github.com/libratbag/libratbag/issues/136#issue-215520871 contains something about creating udev rule, but so far I don't think I was able to configure that properly as running `./ratbag-command` without sudo prints _ratbag error: SteelSeries Sensei Raw Gaming Mouse: error opening hidraw node (Permission denied)_,
- setting up D-Bus parts - when I run e.g. `sudo ./ratbagctl.devel --keepalive list` I can see `org.freedesktop.ratbag_devel1.conf` copied to `/etc/dbus-1/system.d` but I think there are some additional pieces besides this file to get `ratbagd` run as a non-root user - https://github.com/libratbag/libratbag/issues/421#issuecomment-345186760 contains something about it, but no details really.

(I know that `ratbag-command` does not use `ratbagd` over D-Bus but uses `libratbag` directly.)

If such approach as desired by me isn't really possible then I can manage probably with one of the ways described in https://github.com/libratbag/libratbag/issues/421#issuecomment-345186760. I would just need to learn how to connect to the debugger from the IDE (Gnome Builder or whatever else will work). I understand that overall this falls under 'learn how to develop under Linux, really'. I remember that I managed to set up this stuff long time ago when I contributed 2 or 3 meager UI fixes to Nautilus. But some docs that push you into the proper directions can be very helpful.

**Comment by FFY00** [comment]
`ratbagctl.devel` already spawns a temporary `ratbagd` instance, that's why it takes so long compared to `ratbagctl`. You can just call it in your IDE, no need to mess with ratbagd, dbus or anything like that.

**Comment by whot** [comment]
simplest udev rules to control the hidraw devices would be:
```
KERNEL=="hidraw*", MODE="0666"
KERNEL=="hidraw*", GROUP="input"
```
either one of those should do, in the second case you'll have to add yourself to the `input` group. Just be aware of all the side-effects here, the first line would give rw acess to all users on all input devices. Perfect for a keyboard sniffer. Either way, with that in place, `ratbag-command` should work.

As for DBus, I think this could work, but I haven't tried it.
```
-        <policy user="root">
+       <policy context="default">
```

**Comment by hipnoizz** [comment]
Neither of ways described in https://github.com/libratbag/libratbag/issues/421#issuecomment-345186760 worked for me.

`ratbagctl.devel --keepalive` does not work, I need to add `sudo`. Then an instance of `ratbagd.devel` is started, but when I tried to attach my IDE to the process is says _ptrace: Operation not permitted._ Apparently I beed to dig deeper to get this working.

`python3 ratbag-command` needs `sudo` as well. It's just a Python script anyway which does not spawn any 'normal' process (I think...), so I'm not sure what should I attach my IDE to.

Actually after some more tinkering I get running/debugging `ratbagd` from IDE for non-root user. To do that:
- I added udev rule based on https://github.com/libratbag/libratbag/issues/136#issue-215520871 - need some invocation of `udevadm` to reload rules, but it makes `ratbag-command` work for non-root users as well,
- copied `org.freedesktop.ratbag1.conf` to `/etc/dbus-1/system.d` and changed user from `root` to my own.

I understand that it would probably clash with ratbag installed from a repository (PPA under Ubuntu) - that's why *.devel executables add some hash-based suffix to service name. Anyway at the moment I can run and debug `ratbagd` from my IDE and then use `ratbagctl` to interact with it.

**Comment by whot** [comment]
`ratbatctl.devel` installs a custom dbus policy so we can start test versions of ratbagd without interfering with the system one. Unless you `chmod /etc/dbus-1/system.d` you will fail just because of that.

`ptrace` - is this [selinux](https://fedoraproject.org/wiki/Features/SELinuxDenyPtrace) or some related issue?

`ratbag-command` needs root to access `/dev/hidraw` devices. And yes, the frontend is a python script so we can re-use the same cmdline arguments with `ratbagctl`.

**Comment by hipnoizz** [comment]
I've documented a way to run `ratbagd` as a non-root user here:https://github.com/hipnoizz/libratbag/wiki/Running-ratbagd-As-Non-root-In-Development. It could copied to libratbag Wiki if someone is interested. Either way we can close this ticket I think.

**Comment by bentiss** [comment]
Thanks for the documentation. However, I am not so sure we want to put it as it is in the wiki. You are changing the hidraw node permissions of you mouse to `0666`. But if not taken we care, there is a high chance users will inadvertently change the hidraw permissions of their keyboard, allowing any userspace program to register the keys as they are pressed.

A big enough warning should at least be present, and it needs to clearly asses the problem of enabling this. In your case, I agree that adding `0666` to your mouse is not ideal but not *that* problematic.

**Comment by whot** [comment]
Might be a better option to copy what the `/dev/input` devices do and assign a specific group to them (e.g. `input` or a newly created `hidraw`), then add yourself to that group. Ever so slightly safer than just using `0666`, but anything running as your user can still access it then, so...
